src/main.rs

```diff
@@ -36,7 +36,7 @@ fn main() -> CliResult {
 }
 
 /// Decide which commands to run given the input from the shell
-fn run_floki_from_args(args: &Cli) -> Result<(), Error> {
+fn run_floki_from_args(args: Cli) -> Result<(), Error> {
```

Comment:
move entails a `memmove` under the covers, although it's possible the compiler is smart enough to elide it (it's on the stack already, and doesn't outlive the creating frame). Passing by reference guarantees there is no copy.

Aside: @maxdymond I made the argument move should be cheap before (against a comment you made). That was incorrect (I ought to go back and fix it).

Reply:
I did it to avoid the need to clone `shell` when passing it to [`gen_completions_to`](https://docs.rs/clap/2.33.1/clap/struct.App.html#method.gen_completions_to), but actually it looks like [`Shell`]() implements `Copy`, so we can deref it instead.

Reply:
@rlupton20 I looked into this a little, and spoke to @bossmc 

In `release` mode this won't do a memcpy (in `debug` it will) I made a little model: see the difference in LLVM IR & ASM between debug and release modes in: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1b2b5580885e556532c0cfa34c19cda3

This is good because as the caller it's good to be able to hand over ownership of data that you're done with. It also can avoid the need for clones in the function being called.

Reply:
in this case `clap::Shell` implements `Copy` so it ends up being identical either way.

Reply:
Nice, good to know. Its interesting that it passes a pointer to the structure, rather than just resolving it as `[rsp+0x8]`.

I'm guessing the way moves work in Rust it can't guarantee copy elision in general, but I wonder if it can in the case of e.g. nesting stack frames.